src/m365/spo/commands/file/file-roleassignment-remove.ts

```diff
@@ -0,0 +1,211 @@
+
+interface CommandArgs {
+  options: Options;
+}
+
+export interface Options extends GlobalOptions {
+  webUrl: string;
+  fileUrl?: string;
+  fileId?: string;
+  principalId?: number;
+  upn?: string;
+  groupName?: string;
+  confirm?: boolean;
+}
+
+class SpoFileRoleAssignmentRemoveCommand extends SpoCommand {
+  public get name(): string {
+    return commands.FILE_ROLEASSIGNMENT_REMOVE;
+  }
+
+  public get description(): string {
+    return 'Removes role assignment from a file';
+  }
+
+  constructor() {
+    super();
+
+    this.#initTelemetry();
+    this.#initOptions();
+    this.#initValidators();
+    this.#initOptionSets();
+  }
+
+  #initTelemetry(): void {
+    this.telemetry.push((args: CommandArgs) => {
+      Object.assign(this.telemetryProperties, {
+        fileUrl: typeof args.options.fileUrl !== 'undefined',
+        fileId: typeof args.options.fileId !== 'undefined',
+        principalId: typeof args.options.principalId !== 'undefined',
+        upn: typeof args.options.upn !== 'undefined',
+        groupName: typeof args.options.groupName !== 'undefined',
+        confirm: (!(!args.options.confirm)).toString()
+      });
+    });
+  }
+
+  #initOptions(): void {
+    this.options.unshift(
+      {
+        option: '-u, --webUrl <webUrl>'
+      },
+      {
+        option: '--fileUrl [fileUrl]'
+      },
+      {
+        option: '-i, --fileId [fileId]'
+      },
+      {
+        option: '--principalId [principalId]'
+      },
+      {
+        option: '--upn [upn]'
+      },
+      {
+        option: '--groupName [groupName]'
+      },
+      {
+        option: '--confirm'
+      }
+    );
+  }
+
+  #initValidators(): void {
+    this.validators.push(
+      async (args: CommandArgs) => {
+        const isValidSharePointUrl: boolean | string = validation.isValidSharePointUrl(args.options.webUrl);
+        if (isValidSharePointUrl !== true) {
+          return isValidSharePointUrl;
+        }
+
+        if (args.options.principalId && isNaN(args.options.principalId)) {
+          return `Specified principalId ${args.options.principalId} is not a number`;
+        }
+
+        if (args.options.fileId && !validation.isValidGuid(args.options.fileId)) {
+          return `${args.options.fileId} is not a valid GUID`;
+        }
+
+        return true;
+      }
+    );
+  }
+
+  #initOptionSets(): void {
+    this.optionSets.push(
+      ['fileUrl', 'fileId'],
+      ['upn', 'groupName', 'principalId']
+    );
+  }
+
+  public async commandAction(logger: Logger, args: CommandArgs): Promise<void> {
+    const removeRoleAssignment: () => Promise<void> = async (): Promise<void> => {
+      try {
+        const fileURL: string = await this.getFileURL(args);
+
+        if (args.options.groupName) {
+          args.options.principalId = await this.GetGroupPrincipalId(args.options);
+        }
+        else if (args.options.upn) {
+          args.options.principalId = await this.GetUserPrincipalId(args.options);
+        }
```

Comment:
We advise against setting the command options yourself. Command options should be treated as read-only properties.
Here we should introduce a new local variable and pass it to other methods

```suggestion
        let principalId: number;
        if (args.options.groupName) {
          principalId = await this.GetGroupPrincipalId(args.options);
        }
        else if (args.options.upn) {
          principalId = await this.GetUserPrincipalId(args.options);
        }
```